File: mobile/novalang_flutter/lib/core/utils/level_display.dart
import 'level_display_labels.dart';

/// Display name for a level code.
///
/// Labels are loaded from `assets/shared/level_labels.json`
/// (mirrored from `shared/i18n/level_labels.json`).
///
/// [nativeLanguage] is used for Japanese + Vietnamese labels,
/// matching Web `getLevelDisplayName()` behavior.
String getLevelDisplayName(
  String levelCode,
  String learningLanguage, {
  String? nativeLanguage,
}) {
  return LevelDisplayLabels.instance.displayName(
    levelCode,
    learningLanguage,
    nativeLanguage: nativeLanguage,
  );
}

List<String> get levelOrder => LevelDisplayLabels.instance.levelOrder;

File: shared/levelDisplay.ts
import type { LanguageCode, LevelId, NativeLanguageCode } from "./types.js";
import levelLabelsConfig from "./i18n/level_labels.json" with { type: "json" };

type LevelLabelsFile = {
  levelOrder: LevelId[];
  tracks: Record<
    string,
    { labels: Record<string, Partial<Record<LevelId, string>>> }
  >;
};

const config = levelLabelsConfig as LevelLabelsFile;

/** Stable level order — sourced from shared/i18n/level_labels.json */
export const levelOrder: LevelId[] = config.levelOrder;

const defaultLabels = config.tracks.default?.labels.en ?? {};
const japaneseLabelsEn = config.tracks.ja?.labels.en ?? {};
const japaneseLabelsVi = config.tracks.ja?.labels.vi ?? {};

export function getLevelDisplayName(
  levelCode: LevelId,
  learningLanguage: LanguageCode,
  nativeLanguage?: NativeLanguageCode,
): string {
  if (learningLanguage === "ja") {
    const labels =
      nativeLanguage === "vi" ? japaneseLabelsVi : japaneseLabelsEn;
    return labels[levelCode] ?? levelCode;
  }
  return defaultLabels[levelCode] ?? levelCode;
}
